File: medium/palin/palin.py
#!/usr/bin/python

cases = int(raw_input())
def is_palin(num) :
		if num<10 :
			return 0
		else :
			num_str = str(num)
			l = len(num_str)
			for i in range ( 0 , l/2 ) :
				if num_str[i] != num_str[-1-i] :
					return 0
		return 1 

while cases :
	num = int(raw_input())
	if is_palin(num) : 
		num+=1
	num_str = str(num)
	#check if alreay palindrome in that case increment 
	l = len(num_str)
	if l == 1 : #single string 
		if num == 9 :
			print "11" 
		else :
		  	print num+1 

		continue	

	if l %2 == 0 : #even
		left = num_str[:l/2]
		right = num_str[l/2:]
		r_num = int(right)
		l_num = int(left)
		lr_num = int(left[::-1])
		if lr_num > r_num :
			ans = left + left[::-1] 
		elif lr_num < r_num :
			l_num+=1;
			left = str(l_num)
			ans = left + left[::-1]

	else : 
		center  = int(num_str[l/2])
		left = num_str[:l/2]
		right = num_str[l/2+1:]
		l_num= int(left)
		r_num = int(right)
		lr_num = int(left[::-1])
		if lr_num > r_num :
			ans = left + str(center) + left[::-1]
		elif lr_num < r_num :
			carry = 0 
			center+=1
			if center == 10 : 
				center = 0 
				carry = 1 
			left = str( l_num+carry )
			ans = left + str(center) + left[::-1]

	
	print ans 
	cases-=1

File: medium/palin/palin.cpp
#include<iostream>
#include<string>
using namespace std;


int main()
{
	int cases;
	cin>>cases;
	while(cases--)
	{
		string num_str ;
		cin>>num_str;
		long long len  = num_str.size();
		long long j,k;
		string ans ;
		j=len/2-1,k=len/2;
		if(len%2 == 0 )
		{
			while( num_str[j] == num_str[k] && j>=0 )
			{
				j--,k++ ;
			}
			if(num_str[j] > num_str[k] )
			{
				//make left = right 
				ans.append(num_str,0,len/2);
				ans.append(num_str.rbegin()+len/2,num_str.rend());
				
			}
			else //increment left part 
			{
				int carry = 1 ;
				for(long long i = len/2-1; i>=0 ; i--)
				{
					char tmp ;
					tmp = num_str[i] + carry ;
				        if ( tmp > '9' )
					{
						tmp='0';carry=1 ;
						num_str[i] = tmp ;
					}
					else
					{
						carry=0;
						num_str[i] = tmp ;
						break;
					}
				}	
				if(carry)
				{
					ans.push_back('1');
				}
				ans.append(num_str.begin() , num_str.begin()+len/2);
			
				if(carry)
				{
					ans.append(num_str.rbegin()+len/2+1,num_str.rend());
					ans.push_back('1');
				}
				else 
				{
					ans.append(num_str.rbegin()+len/2,num_str.rend());
				}
			}
		}
		else //odd  length 
		{
			k++;
			while( num_str[j] == num_str[k] && j>=0 )
			{
				j--,k++ ;
			}
			if( num_str[j] > num_str[k] )
			{
				ans.append(num_str,0,len/2);
				ans.push_back(num_str[len/2]);
				ans.append(num_str.rbegin()+len/2+1,num_str.rend()); 
			}
			else
			{

				char cen = num_str[len/2];
				cen++;
				int carry = 0 ;
				if ( cen > '9' )
				{
					cen='0';
					carry = 1 ;
				}
				if(carry) 
				{
					for(long long i = len/2-1; i>=0 ; i--)
					{
						char tmp ;
						tmp = num_str[i] + carry ;
						if ( tmp > '9' )
						{
							tmp='0';carry=1 ;
							num_str[i] = tmp ;
						}
						else
						{
							carry=0;
							num_str[i] = tmp ;
							break;
						}
					}
				}
				if(carry)
				{
					ans.push_back('1');
				}
				ans.append(num_str.begin() , num_str.begin()+len/2);
				ans.push_back(cen);	
				if(carry)
				{
					ans.append(num_str.rbegin()+len/2+2,num_str.rend());
					ans.push_back('1');
				}
				else 
				{
					ans.append(num_str.rbegin()+len/2+1,num_str.rend());
				}
			   }
		}
				

		cout<<ans<<endl;

	}
}

